File: turn-orchestrator/src/transitions.rs
//! One-shot transition dispatcher. Drives `record` forward by exactly one
//! state, then returns. Callers persist the new record and decide whether
//! to re-publish `turn::step_requested`.

use iii_sdk::III;

use crate::state::{TurnState, TurnStateRecord};
use crate::states;

pub async fn step(iii: &III, record: &mut TurnStateRecord) -> anyhow::Result<()> {
    match record.state {
        TurnState::Provisioning => states::handle_provisioning(iii, record).await?,
        TurnState::AwaitingAssistant => states::handle_awaiting(iii, record).await?,
        TurnState::AssistantStreaming => states::handle_streaming(iii, record).await?,
        TurnState::AssistantFinished => states::handle_finished(iii, record).await?,
        TurnState::ToolPrepare => states::handle_prepare(iii, record).await?,
        TurnState::ToolExecute => states::handle_execute(iii, record).await?,
        TurnState::ToolFinalize => states::handle_finalize(iii, record).await?,
        TurnState::SteeringCheck => states::handle_steering(iii, record).await?,
        TurnState::TearingDown => states::handle_tearing_down(iii, record).await?,
        TurnState::Stopped => {
            // No-op. Idempotent terminal state.
        }
    }
    Ok(())
}

File: harness-node/src/turn-orchestrator/transitions.ts
import type { ISdk } from '../runtime/iii.js';
import type { TurnOrchestratorConfig } from './config.js';
import type { TurnStateRecord } from './state.js';
import { handleAwaiting, handleFinished, handleStreaming } from './states/assistant.js';
import {
  handleAwaitingApproval,
  handleExecute,
  handleFinalize,
  handlePrepare,
} from './states/functions.js';
import { handleProvisioning } from './states/provisioning.js';
import { handleSteering } from './states/steering.js';
import { handleTearingDown } from './states/tearing-down.js';

export async function step(
  iii: ISdk,
  cfg: TurnOrchestratorConfig,
  rec: TurnStateRecord,
): Promise<void> {
  switch (rec.state) {
    case 'provisioning':
      return handleProvisioning(iii, cfg, rec);
    case 'awaiting_assistant':
      return handleAwaiting(iii, rec);
    case 'assistant_streaming':
      return handleStreaming(iii, rec);
    case 'assistant_finished':
      return handleFinished(iii, rec);
    case 'function_prepare':
      return handlePrepare(iii, rec);
    case 'function_execute':
      return handleExecute(iii, cfg, rec);
    case 'function_awaiting_approval':
      return handleAwaitingApproval(iii, rec);
    case 'function_finalize':
      return handleFinalize(iii, rec);
    case 'steering_check':
      return handleSteering(iii, rec);
    case 'tearing_down':
      return handleTearingDown(iii, rec);
    case 'stopped':
      return; // idempotent terminal
  }
}
